docs_src/src/guide/configuration.md
---
meta:
  - name: "og:title"
    content: "Configuration - Goyave"
  - name: "twitter:title"
    content: "Configuration - Goyave"
  - name: "title"
    content: "Configuration - Goyave"
---

# Configuration

[[toc]]

## Introduction


The Goyave framework lets you configure its core and your application.
To configure your application, use the `config.json` file at your project's root. If you are using the template project, copy `config.example.json` to `config.json`. `config.json` should be ignored in your `.gitignore` file as it can differ from one developer to another. To avoid accidental commit or change to the default project's config, it is a good practice to ignore this file and define the project's default config in `config.example.json`.

If this config file misses some config entries, the default values will be used. Refer to the [configuration reference](#configuration-reference) to know more. 

All entries are **validated**. That means that the application will not start if you provided an invalid value in your config (for example if the specified port is not a number). That also means that a goroutine trying to change a config entry with the incorrect type will panic.  
Entries can be registered with a default value, their type and authorized values from any package. 

Configuration can be used concurrently safely.

The following JSON file is an example of default configuration:

```json
{
  "app": {
    "name": "goyave_template",
    "environment": "localhost",
    "debug": true,
    "defaultLanguage": "en-US"
  },
  "server": {
    "host": "127.0.0.1",
    "maintenance": false,
    "protocol": "http",
    "domain": "",
    "port": 8080,
    "httpsPort": 8081,
    "timeout": 10,
    "maxUploadSize": 10,
    "tls": {
      "cert": "/path/to/cert",
      "key": "/path/to/key"
    }
  },
  "database": {
    "connection": "mysql",
    "host": "127.0.0.1",
    "port": 3306,
    "name": "goyave",
    "username": "root",
    "password": "root",
    "options": "charset=utf8mb4&collation=utf8mb4_general_ci&parseTime=true&loc=Local",
    "maxOpenConnections": 20,
    "maxIdleConnections": 20,
    "maxLifetime": 300,
    "autoMigrate": false
  }
}
```

## Terminology

**Entry**: a configuration entry is a value accessible using a key.

**Registering an entry**: informs the framework that an entry with the given key is expected. Registering an entry allows to set a default value to be used if this entry is not provided in an app's configuration file, to enforce a certain type for this entry (for example if it needs to be an integer), and to set a list of allowed values.

**Category**: a category is represented by a JSON object in your configuration file, delimited by braces. Sub-categories are categories that are not at root level, for example: `server.tls` is a sub-category of the `server` category.

## Environment configuration

Most projects need different configuration values based on the environment. For example, you won't connect to the same database if you're in local development, in a testing environment inside continuous integration pipelines, or in production. Goyave supports multiple configurations and will pick the appropriate one depending on the environment variable `GOYAVE_ENV`.

Since `v2.0.0`, you can use custom environments.

| GOYAVE_ENV                    | Config file              |
|-------------------------------|--------------------------|
| test                          | `config.test.json`       |
| production                    | `config.production.json` |
| *custom_env*                  | `config.custom_env.json` |
| local / localhost / *not set* | `config.json`            |

## Using the configuration

Before being able to use the config, import the config package:
``` go
import "github.com/System-Glitch/goyave/v3/config"
```

The configuration is loaded automatically when the server starts, but you can reload it manually if needed.

When the configuration is loaded, all default values are copied to the newly created map holding the configuration. Then, the configuration file is read, parsed and is applied over. This means that all entries from the file override the default ones. However, if an entry has a default value and the same entry is not present in the configuration file, then it is kept as it is. On the other hand, if an entry is present in the configuration file and not in the default values (meaning that this entry is not expected), a new entry will be registered. This new entry will be subsequently validated using the type of its initial value and have an empty slice as authorized values (meaning it can have any value of its type).

The following cases will raise errors when the configuration is being overridden:
- When the configuration file overrides an entry with a category
- When the configuration file overrides a category with an entry

#### config.Load

| Parameters | Return  |
|------------|---------|
|            | `error` |

**Example:**
``` go
config.Load()
```

#### config.LoadFrom

You may need to load a configuration file from a custom path instead of using the standard one. `LoadFrom` lets you load a config file from the given path.

| Parameters    | Return  |
|---------------|---------|
| `path string` | `error` |

**Example:** load the config from the path given through a flag
``` go
import (
  "flag"
  "os"

  "github.com/System-Glitch/goyave/v3"
  "github.com/System-Glitch/goyave/v3/config"
  
  //...
)

func handleFlags() {
	flag.Usage = func() {
		goyave.ErrLogger.Println("usage: " + os.Args[0] + " -config=[config]")
		flag.PrintDefaults()
		os.Exit(1)
	}

	flag.String("config", "", "JSON config file")
	flag.Parse()

	configDir := flag.Lookup("config")
	path := configDir.Value.String()
	if path != configDir.DefValue {
		if err := config.LoadFrom(path); err != nil {
			goyave.ErrLogger.Println(err)
			os.Exit(goyave.ExitInvalidConfig)
		}
	}
}

func main() {
	handleFlags()

	if err := goyave.Start(route.Register); err != nil {
		os.Exit(err.(*goyave.Error).ExitCode)
	}
}
```

### Using environment variables

<p><Badge text="Since v3.0.0"/></p>

You can use environment variables in your configuration file. Environment variables are identified by the following syntax: `${VARIABLE_NAME}`.

```json
{
  "database": {
    "host": "${DB_HOST}"
  }
}
```

**Note:** *This syntax is strict. If the string doesn't start with `${` or doesn't end with `}`, it will not be considered an environment variable.*

`int`, `float64` and `bool` values are supported. If the configuration entry is expected to be of one of these types, the content of the environment variable will be automatically converted. If the conversion fails, a configuration loading error will be returned.

If an environment variable mentioned in a configuration file is not set, the configuration validation will not pass. Environment variables are not supported inside slices.

### Getting a value

All entries are accessible using **dot-separated paths**. If you want to access the `name` entry in the `app` category, the key will be `app.name`.

::: table
[Get](#config-get)
[GetString](#config-getstring)
[GetBool](#config-getbool)
[GetInt](#config-getint)
[GetFloat](#config-getfloat)
[GetStringSlice](#config-getstringslice)
[GetBoolSlice](#config-getboolslice)
[GetIntSlice](#config-getintslice)
[GetFloatSlice](#config-getfloatslice)
[Has](#config-has)
:::

#### config.Get

Get a generic config entry. 

Prefer using the `GetString`, `GetBool`, `GetInt`, `GetFloat`, ... accessors. If you need a type not covered by those accessors, use `config.Get`. You may need to type-assert the returned value before using it. You can do so safely as the config values and types are validated.

Panics if the entry doesn't exist.

| Parameters   | Return                 |
|--------------|------------------------|
| `key string` | `interface{}` or panic |

**Example:**
``` go
config.Get("app.name") // "goyave"
```

#### config.GetString

Get a string config entry. Panics if the entry doesn't exist or is not a `string`.

| Parameters   | Return            |
|--------------|-------------------|
| `key string` | `string` or panic |

**Example:**
``` go
config.GetString("server.protocol") // "http"
```

#### config.GetBool

Get a bool config entry. Panics if the entry doesn't exist or is not a `bool`.

| Parameters   | Return          |
|--------------|-----------------|
| `key string` | `bool` or panic |

**Example:**
``` go
config.GetBool("app.debug") // true
```

#### config.GetInt

<p><Badge text="Since v3.0.0"/></p>

Get an int config entry. Panics if the entry doesn't exist or is not an `int`.

| Parameters   | Return         |
|--------------|----------------|
| `key string` | `int` or panic |

**Example:**
``` go
config.GetInt("server.port") // 8080
```

#### config.GetFloat

<p><Badge text="Since v3.0.0"/></p>

Get a float config entry. Panics if the entry doesn't exist or is not a `float64`.

| Parameters   | Return             |
|--------------|--------------------|
| `key string` | `float64` or panic |

**Example:**
``` go
config.GetInt("server.port") // 8080
```

#### config.GetStringSlice

<p><Badge text="Since v3.0.0"/></p>

Get a string slice config entry. Panics if the entry doesn't exist or is not a `[]string`.

| Parameters   | Return              |
|--------------|---------------------|
| `key string` | `[]string` or panic |

**Example:**
``` go
config.GetStringSlice("stringSlice") // [val1 val2]
```

#### config.GetBoolSlice

<p><Badge text="Since v3.0.0"/></p>

Get a bool slice config entry. Panics if the entry doesn't exist or is not a `[]bool`.

| Parameters   | Return            |
|--------------|-------------------|
| `key string` | `[]bool` or panic |

**Example:**
``` go
config.GetBoolSlice("boolSlice") // [true false]
```

#### config.GetIntSlice

<p><Badge text="Since v3.0.0"/></p>

Get an int slice config entry. Panics if the entry doesn't exist or is not a `[]int`.

| Parameters   | Return           |
|--------------|------------------|
| `key string` | `[]int` or panic |

**Example:**
``` go
config.GetIntSlice("intSlice") // [3 5]
```

#### config.GetFloatSlice

<p><Badge text="Since v3.0.0"/></p>

Get a float64 slice config entry. Panics if the entry doesn't exist or is not a `[]float64`.

| Parameters   | Return               |
|--------------|----------------------|
| `key string` | `[]float64` or panic |

**Example:**
``` go
config.GetFloatSlice("floatSlice") // [1.42 3.24]
```

#### config.Has

Check if a config entry exists.

| Parameters   | Return |
|--------------|--------|
| `key string` | `bool` |

**Example:**
``` go
config.Has("app.name") // true
```

### Setting a value

You can set a config value at runtime with the `config.Set(key, value)` function. Bear in mind that this change is **temporary** and will be lost after your application restarts or if the config is reloaded. This function is mainly used for testing purposes. Values set using this function are still being validated: your application will panic and revert changes if the validation doesn't pass.

Use `nil` to unset a value.

- A category cannot be replaced with an entry.
- An entry cannot be replaced with a category.
- New categories can be created with they don't already exist.
- New entries can be created if they don't already exist. This new entry will be subsequently validated using the type of its initial value and have an empty slice as authorized values (meaning it can have any value of its type)

#### config.Set

| Parameters          | Return          |
|---------------------|-----------------|
| `key string`        | `void` or panic |
| `value interface{}` |                 |

**Example:**
``` go
config.Set("app.name", "my awesome app")
```

## Custom config entries

<p><Badge text="Since v3.0.0"/></p>

Configuration can be expanded. It is very likely that a plugin or a package you're developing is using some form of options. These options can be added to the configuration system so it is not needed to set them in the code or to make some wiring.

#### config.Register

Register a new config entry and its validation.

Each module should register its config entries in an `init()` function, even if they don't have a default value, in order to ensure they will be validated.

Each module should use its own category and use a name both expressive and unique to avoid collisions. For example, the `auth` package registers, among others, `auth.basic.username` and `auth.jwt.expiry`, thus creating a category for its package, and two subcategories for its features.

To register an entry without a default value (only specify how it will be validated), set `Entry.Value` to `nil`.

Is `IsSlice` is `true`, the value of the entry will be a slice of the given `Type`. Authorized values for slices define the values that can be used inside the slice. It doesn't represent the value of the slice itself (content and order).

Panics if an entry already exists for this key and is not identical to the one passed as parameter of this function. On the other hand, if the entries are identical, no conflict is expected so the configuration is left in its current state.

| Parameters          | Return          |
|---------------------|-----------------|
| `key string`        | `void` or panic |
| `kind config.Entry` |                 |

**Example:**
``` go
func init() {
  config.Register("my-plugin.name", config.Entry{
    Value:            "default value",
    Type:             reflect.String,
    IsSlice:          false,
    AuthorizedValues: []interface{}{},
  })
  
  // Without a default value (only validation)
  config.Register("my-plugin.protocol", config.Entry{
    Value:            nil,
    Type:             reflect.String,
    IsSlice:          false,
    AuthorizedValues: []interface{}{"ftp", "sftp", "scp"},
  })

  // Slice
  config.Register("my-plugin.remoteHosts", config.Entry{
    Value:            []string{"first host", "second host"},
    Type:             reflect.String,
    IsSlice:          true,
    AuthorizedValues: []interface{}{},
  })
}
```

## Configuration reference

### App category

| Entry           | Type     | Default     | Note                                                                                                           |
|-----------------|----------|-------------|----------------------------------------------------------------------------------------------------------------|
| name            | `string` | "goyave"    |                                                                                                                |
| environment     | `string` | "localhost" |                                                                                                                |
| debug           | `bool`   | `true`      | When activated, print stacktrace on error and sends error message in response. **Disable this in production!** |
| defaultLanguage | `string` | "en-US"     | See the [Localization](./advanced/localization.html) section                                                   |

### Server category

| Entry         | Type      | Accepted values | Default     | Note                                                                      |
|---------------|-----------|-----------------|-------------|---------------------------------------------------------------------------|
| host          | `string`  | any             | "127.0.0.1" |                                                                           |
| domain        | `string`  | any             | ""          | Used for URL generation. Leave empty to use IP instead.                   |
| protocol      | `string`  | "http", "https" | "http"      | See the [HTTPS](#setting-up-https) section                                |
| port          | `int`     | any             | `8080`      |                                                                           |
| httpsPort     | `int`     | any             | `8081`      |                                                                           |
| timeout       | `int`     | any             | `10`        | Timeout in seconds                                                        |
| maxUploadSize | `float64` | any             | `10`        | Maximum size of the request, in MiB                                       |
| maintenance   | `bool`    | any             | `false`     | If `true`, start the server in maintenance mode. (Always return HTTP 503) |

#### TLS sub-category

| Entry | Type     | Default | Note                  |
|-------|----------|---------|-----------------------|
| cert  | `string` | none    | Path to your TLS cert |
| key   | `string` | none    | Path to your TLS key  |

### Database category

| Entry              | Type     | Default                                                                 | Note                                                      |
|--------------------|----------|-------------------------------------------------------------------------|-----------------------------------------------------------|
| connection         | `string` | "none"                                                                  | See the [Database](./basics/database.html) guide          |
| host               | `string` | "127.0.0.1"                                                             |                                                           |
| port               | `int`    | `3306`                                                                  |                                                           |
| name               | `string` | "goyave"                                                                |                                                           |
| username           | `string` | "root"                                                                  |                                                           |
| password           | `string` | "root"                                                                  |                                                           |

config/config.go
package config

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strconv"
	"strings"
	"sync"

	"github.com/System-Glitch/goyave/v3/helper"
)

type object map[string]interface{}

// Entry is the internal reprensentation of a config entry.
// It contains the entry value, its expected type (for validation)
// and a slice of authorized values (for validation too). If this slice
// is empty, it means any value can be used, provided it is of the correct type.
type Entry struct {
	Value            interface{}
	Type             reflect.Kind
	IsSlice          bool
	AuthorizedValues []interface{} // Leave empty for "any"
}

var config object

var configDefaults object = object{
	"app": object{
		"name":            &Entry{"goyave", reflect.String, false, []interface{}{}},
		"environment":     &Entry{"localhost", reflect.String, false, []interface{}{}},
		"debug":           &Entry{true, reflect.Bool, false, []interface{}{}},
		"defaultLanguage": &Entry{"en-US", reflect.String, false, []interface{}{}},
	},
	"server": object{
		"host":          &Entry{"127.0.0.1", reflect.String, false, []interface{}{}},
		"domain":        &Entry{"", reflect.String, false, []interface{}{}},
		"protocol":      &Entry{"http", reflect.String, false, []interface{}{"http", "https"}},
		"port":          &Entry{8080, reflect.Int, false, []interface{}{}},
		"httpsPort":     &Entry{8081, reflect.Int, false, []interface{}{}},
		"timeout":       &Entry{10, reflect.Int, false, []interface{}{}},
		"maxUploadSize": &Entry{10.0, reflect.Float64, false, []interface{}{}},
		"maintenance":   &Entry{false, reflect.Bool, false, []interface{}{}},
		"tls": object{
			"cert": &Entry{nil, reflect.String, false, []interface{}{}},
			"key":  &Entry{nil, reflect.String, false, []interface{}{}},
		},
	},
	"database": object{
		"connection":         &Entry{"none", reflect.String, false, []interface{}{}},
		"host":               &Entry{"127.0.0.1", reflect.String, false, []interface{}{}},
		"port":               &Entry{3306, reflect.Int, false, []interface{}{}},
		"name":               &Entry{"goyave", reflect.String, false, []interface{}{}},
		"username":           &Entry{"root", reflect.String, false, []interface{}{}},
		"password":           &Entry{"root", reflect.String, false, []interface{}{}},
		"options":            &Entry{"charset=utf8mb4&collation=utf8mb4_general_ci&parseTime=true&loc=Local", reflect.String, false, []interface{}{}},
		"maxOpenConnections": &Entry{20, reflect.Int, false, []interface{}{}},
		"maxIdleConnections": &Entry{20, reflect.Int, false, []interface{}{}},
		"maxLifetime":        &Entry{300, reflect.Int, false, []interface{}{}},
		"autoMigrate":        &Entry{false, reflect.Bool, false, []interface{}{}},
	},
}

var mutex = &sync.RWMutex{}

// Register a new config entry and its validation.
//
// Each module should register its config entries in an "init()"
// function, even if they don't have a default value, in order to
// ensure they will be validated.
// Each module should use its own category and use a name both expressive
// and unique to avoid collisions.
// For example, the "auth" package registers, among others, "auth.basic.username"
// and "auth.jwt.expiry", thus creating a category for its package, and two subcategories
// for its features.
//
// To register an entry without a default value (only specify how it
// will be validated), set "Entry.Value" to "nil".
//
// Panics if an entry already exists for this key and is not identical to the
// one passed as parameter of this function. On the other hand, if the entries
// are identical, no conflict is expected so the configuration is left in its
// current state.
func Register(key string, entry Entry) {
	mutex.Lock()
	defer mutex.Unlock()
	category, entryKey, exists := walk(configDefaults, key)
	if exists {
		if !reflect.DeepEqual(&entry, category[entryKey].(*Entry)) {
			panic(fmt.Sprintf("Attempted to override registered config entry %q", key))
		}
	} else {
		category[entryKey] = &entry
	}
}

// Load loads the config.json file in the current working directory.
// If the "GOYAVE_ENV" env variable is set, the config file will be picked like so:
// - "production": "config.production.json"
// - "test": "config.test.json"
// - By default: "config.json"
func Load() error {
	return LoadFrom(getConfigFilePath())
}

// LoadFrom loads a config file from the given path.
func LoadFrom(path string) error {
	mutex.Lock()
	defer mutex.Unlock()
	config = make(object, len(configDefaults))
	loadDefaults(configDefaults, config)

	conf, err := readConfigFile(path)
	if err != nil {
		config = nil
		return err
	}

	if err := override(conf, config); err != nil {
		config = nil
		return err
	}

	if err := config.validate(""); err != nil {
		config = nil
		return fmt.Errorf("Invalid config:%s", err.Error())
	}

	return nil
}

// IsLoaded returns true if the config have been loaded.
func IsLoaded() bool {
	mutex.RLock()
	defer mutex.RUnlock()
	return config != nil
}

// Clear unloads the config.
// DANGEROUS, should only be used for testing.
func Clear() {
	mutex.Lock()
	config = nil
	mutex.Unlock()
}

// Get a config entry. Panics if the entry doesn't exist.
func Get(key string) interface{} {
	if val, ok := get(key); ok {
		return val
	}

	panic(fmt.Sprintf("Config entry \"%s\" doesn't exist", key))
}

func get(key string) (interface{}, bool) {
	mutex.RLock()
	defer mutex.RUnlock()
	if config == nil {
		panic("Config is not loaded")
	}
	currentCategory := config
	b := 0
	e := strings.Index(key, ".")
	if e == -1 {
		e = len(key)
	}
	for path := key[b:e]; ; path = key[b:e] {
		entry, ok := currentCategory[path]
		if !ok {
			break
		}

		if category, ok := entry.(object); ok {
			currentCategory = category
		} else {
			val := entry.(*Entry).Value
			return val, val != nil // nil means unset
		}

		if e+1 <= len(key) {
			b = e + 1
			newE := strings.Index(key[b:], ".")
			if newE == -1 {
				e = len(key)
			} else {
				e = newE + b
			}
		}
	}
	return nil, false
}

// GetString a config entry as string.
// Panics if entry is not a string or if it doesn't exist.
func GetString(key string) string {
	str, ok := Get(key).(string)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a string", key))
	}
	return str
}

// GetBool a config entry as bool.
// Panics if entry is not a bool or if it doesn't exist.
func GetBool(key string) bool {
	val, ok := Get(key).(bool)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a bool", key))
	}
	return val
}

// GetInt a config entry as int.
// Panics if entry is not an int or if it doesn't exist.
func GetInt(key string) int {
	val, ok := Get(key).(int)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not an int", key))
	}
	return val
}

// GetFloat a config entry as float64.
// Panics if entry is not a float64 or if it doesn't exist.
func GetFloat(key string) float64 {
	val, ok := Get(key).(float64)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a float64", key))
	}
	return val
}

// GetStringSlice a config entry as []string.
// Panics if entry is not a string slice or if it doesn't exist.
func GetStringSlice(key string) []string {
	str, ok := Get(key).([]string)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a string slice", key))
	}
	return str
}

// GetBoolSlice a config entry as []bool.
// Panics if entry is not a bool slice or if it doesn't exist.
func GetBoolSlice(key string) []bool {
	str, ok := Get(key).([]bool)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a bool slice", key))
	}
	return str
}

// GetIntSlice a config entry as []int.
// Panics if entry is not an int slice or if it doesn't exist.
func GetIntSlice(key string) []int {
	str, ok := Get(key).([]int)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not an int slice", key))
	}
	return str
}

// GetFloatSlice a config entry as []float64.
// Panics if entry is not a float slice or if it doesn't exist.
func GetFloatSlice(key string) []float64 {
	str, ok := Get(key).([]float64)
	if !ok {
		panic(fmt.Sprintf("Config entry \"%s\" is not a float64 slice", key))
	}
	return str
}

// Has check if a config entry exists.
func Has(key string) bool {
	_, ok := get(key)
	return ok
}

// Set a config entry.
// The change is temporary and will not be saved for next boot.
// Use "nil" to unset a value.
//
//  - A category cannot be replaced with an entry.
//  - An entry cannot be replaced with a category.
//  - New categories can be created with they don't already exist.
//  - New entries can be created if they don't already exist. This new entry
//    will be subsequently validated using the type of its initial value and
//    have an empty slice as authorized values (meaning it can have any value of its type)
//
// Panics and revert changes in case of error.
func Set(key string, value interface{}) {
	mutex.Lock()
	defer mutex.Unlock()
	if config == nil {
		panic("Config is not loaded")
	}
	category, entryKey, exists := walk(config, key)
	if exists {
		entry := category[entryKey].(*Entry)
		previous := entry.Value
		entry.Value = value
		if err := entry.validate(key); err != nil {
			entry.Value = previous
			panic(err)
		}
		category[entryKey] = entry
	} else {
		category[entryKey] = makeEntryFromValue(value)
	}
}

// walk the config using the key. Returns the deepest category, the entry key
// with its path stripped ("app.name" -> "name") and true if the entry already
// exists, false if it's not registered.
func walk(currentCategory object, key string) (object, string, bool) {
	if key == "" {
		panic("Empty key is not allowed")
	}

	if key[len(key)-1:] == "." {
		panic("Keys ending with a dot are not allowed")
	}

	b := 0
	e := strings.Index(key, ".")
	if e == -1 {
		e = len(key)
	}
	for catKey := key[b:e]; ; catKey = key[b:e] {
		entry, ok := currentCategory[catKey]
		if !ok {
			// If categories are missing, create them
			currentCategory = createMissingCategories(currentCategory, key[b:])
			i := strings.LastIndex(key, ".")
			if i == -1 {
				catKey = key
			} else {
				catKey = key[i+1:]
			}

			// Entry doesn't exist and is not registered
			return currentCategory, catKey, false
		}

		if category, ok := entry.(object); ok {
			currentCategory = category
		} else {
			if e < len(key) {
				panic(fmt.Sprintf("Attempted to add an entry to non-category %q", key[:e]))
			}

			// Entry exists
			return currentCategory, catKey, true
		}

		if e+1 <= len(key) {
			b = e + 1
			newE := strings.Index(key[b:], ".")
			if newE == -1 {
				e = len(key)
			} else {
				e = newE + b
			}
		} else {
			break
		}
	}

	panic(fmt.Sprintf("Attempted to replace the %q category with an entry", key))
}

// createMissingCategories based on the key path, starting at the given index.
// Doesn't create anything is not needed.
// Returns the deepest category created, or the provided object if nothing has
// been created.
func createMissingCategories(currentCategory object, path string) object {
	b := 0
	e := strings.Index(path, ".")
	if e == -1 {
		return currentCategory
	}
	for catKey := path[b:e]; ; catKey = path[b:e] {
		newCategory := object{}
		currentCategory[catKey] = newCategory
		currentCategory = newCategory

		if e+1 <= len(path) {
			b = e + 1
			newE := strings.Index(path[b:], ".")
			if newE == -1 {
				return currentCategory
			}
			e = newE + b
		}
	}
}

func loadDefaults(src object, dst object) {
	for k, v := range src {
		if obj, ok := v.(object); ok {
			sub := make(object, len(obj))
			loadDefaults(obj, sub)
			dst[k] = sub
		} else {
			entry := v.(*Entry)
			value := entry.Value
			t := reflect.TypeOf(value)
			if t != nil && t.Kind() == reflect.Slice {
				list := reflect.ValueOf(value)
				length := list.Len()
				slice := reflect.MakeSlice(reflect.SliceOf(t.Elem()), 0, length)
				for i := 0; i < length; i++ {
					slice = reflect.Append(slice, list.Index(i))
				}
				value = slice.Interface()
			}
			dst[k] = &Entry{value, entry.Type, entry.IsSlice, entry.AuthorizedValues}
		}
	}
}

func override(src object, dst object) error {
	for k, v := range src {
		if obj, ok := v.(map[string]interface{}); ok {
			if dstObj, ok := dst[k]; !ok {
				dst[k] = make(object, len(obj))
			} else if _, ok := dstObj.(object); !ok {
				// Conflict: destination is not a category
				return fmt.Errorf("Invalid config:\n\t- Cannot override entry %q with a category", k)
			}
			if err := override(obj, dst[k].(object)); err != nil {
				return err
			}
		} else if entry, ok := dst[k]; ok {
			e, ok := entry.(*Entry)
			if !ok {
				// Conflict: override category with an entry
				return fmt.Errorf("Invalid config:\n\t- Cannot override category %q with an entry", k)
			}
			e.Value = v
		} else {
			// If entry doesn't exist (and is not registered),
			// register it with the type of the type given here
			// and "any" authorized values.
			dst[k] = makeEntryFromValue(v)
		}
	}
	return nil
}

func makeEntryFromValue(value interface{}) *Entry {
	isSlice := false
	t := reflect.TypeOf(value)
	kind := t.Kind()
	if kind == reflect.Slice {
		kind = t.Elem().Kind()
		isSlice = true
	}
	return &Entry{value, kind, isSlice, []interface{}{}}
}

func readConfigFile(file string) (object, error) {
	conf := make(object, len(configDefaults))
	configFile, err := os.Open(file)

	if err == nil {
		defer configFile.Close()
		jsonParser := json.NewDecoder(configFile)
		err = jsonParser.Decode(&conf)
	}
	return conf, err
}

func getConfigFilePath() string {
	env := strings.ToLower(os.Getenv("GOYAVE_ENV"))
	if env == "local" || env == "localhost" || env == "" {
		return "config.json"
	}
	return "config." + env + ".json"
}

func (o object) validate(key string) error {
	message := ""
	valid := true
	for k, entry := range o {
		var subKey string
		if key == "" {
			subKey = k
		} else {
			subKey = key + "." + k
		}
		if category, ok := entry.(object); ok {
			if err := category.validate(subKey); err != nil {
				message += err.Error()
				valid = false
			}
		} else if err := entry.(*Entry).validate(subKey); err != nil {
			message += "\n\t- " + err.Error()
			valid = false
		}
	}

	if !valid {
		return fmt.Errorf(message)
	}
	return nil
}

func (e *Entry) validate(key string) error {
	if e.Value == nil { // nil values means unset
		return nil
	}

	if err := e.tryEnvVarConversion(key); err != nil {
		return err
	}

	t := reflect.TypeOf(e.Value)
	kind := t.Kind()
	if e.IsSlice && kind == reflect.Slice {
		kind = t.Elem().Kind()
	}
	if kind != e.Type {
		if !e.tryIntConversion(kind) {
			var message string
			if e.IsSlice {
				message = "%q must be a slice of %s"
			} else {
				message = "%q type must be %s"
			}

			return fmt.Errorf(message, key, e.Type)
		}
		return nil
	}

	if len(e.AuthorizedValues) > 0 {
		if e.IsSlice {
			// Accepted values for slices define the values that can be used inside the slice
			// It doesn't represent the value of the slice itself (content and order)
			list := reflect.ValueOf(e.Value)
			length := list.Len()
			authorizedValuesList := reflect.ValueOf(e.AuthorizedValues)
			for i := 0; i < length; i++ {
				if !e.authorizedValuesContains(authorizedValuesList, list.Index(i).Interface()) {
					return fmt.Errorf("%q elements must have one of the following values: %v", key, e.AuthorizedValues)
				}
			}
		} else if !helper.Contains(e.AuthorizedValues, e.Value) {
			return fmt.Errorf("%q must have one of the following values: %v", key, e.AuthorizedValues)
		}
	}

	return nil
}

// authorizedValuesContains avoids to recreate the reflect.Value of the list for every check
func (e *Entry) authorizedValuesContains(list reflect.Value, value interface{}) bool {
	length := list.Len()
	for i := 0; i < length; i++ {
		if list.Index(i).Interface() == value {
			return true
		}
	}
	return false
}

func (e *Entry) tryIntConversion(kind reflect.Kind) bool {
	if kind == reflect.Float64 && e.Type == reflect.Int {
		if e.IsSlice {
			return e.convertIntSlice()
		}

		intVal, ok := e.convertInt(e.Value.(float64))
		if ok {
			e.Value = intVal
			return true
		}
	}

	return false
}

func (e *Entry) convertInt(value float64) (int, bool) {
	intVal := int(value)
	if value == float64(intVal) {
		return intVal, true
	}
	return 0, false
}

func (e *Entry) convertIntSlice() bool {
	original := e.Value.([]float64)
	slice := make([]int, len(original))
	for k, v := range original {
		intVal, ok := e.convertInt(v)
		if !ok {
			return false
		}
		slice[k] = intVal
	}
	e.Value = slice
	return true
}

func (e *Entry) tryEnvVarConversion(key string) error {
	str, ok := e.Value.(string)
	if ok {
		val, err := e.convertEnvVar(str, key)
		if err == nil && val != nil {
			e.Value = val
		}
		return err
	}

	return nil
}

func (e *Entry) convertEnvVar(str, key string) (interface{}, error) {
	if strings.HasPrefix(str, "${") && strings.HasSuffix(str, "}") {
		varName := str[2 : len(str)-1]
		value, set := os.LookupEnv(varName)
		if !set {
			return nil, fmt.Errorf("%q: %q environment variable is not set", key, varName)
		}

		switch e.Type {
		case reflect.Int:
			if i, err := strconv.Atoi(value); err == nil {
				return i, nil
			}
			return nil, fmt.Errorf("%q could not be converted to int from environment variable %q of value %q", key, varName, value)
		case reflect.Float64:
			if f, err := strconv.ParseFloat(value, 64); err == nil {
				return f, nil
			}
			return nil, fmt.Errorf("%q could not be converted to float64 from environment variable %q of value %q", key, varName, value)
		case reflect.Bool:
			if b, err := strconv.ParseBool(value); err == nil {
				return b, nil
			}
			return nil, fmt.Errorf("%q could not be converted to bool from environment variable %q of value %q", key, varName, value)
		default:
			// Keep value as string if type is not supported and let validation do its job
			return value, nil
		}
	}

	return nil, nil
}
